en/concept/architecture.md

```diff
@@ -113,34 +113,27 @@ considerably slower.
 The message update rates can be [inspected](../middleware/uorb.md)
 in real-time on the system by running `uorb top`.
 
-## Runtime Environment
+## Runtime Environment {#runtime-environment}
 
-PX4 runs on various operating systems that provide a POSIX-API
-(such as Linux, macOS, NuttX or QuRT). It should also have some form of
-real-time scheduling (e.g. FIFO).
+PX4 runs on various operating systems that provide a POSIX-API (such as Linux, macOS, NuttX or QuRT).
+It should also have some form of real-time scheduling (e.g. FIFO).
 
-The inter-module communication (using [uORB](../middleware/uorb.md)) is based on shared memory. 
-The whole PX4 middleware runs in a single address space, i.e. memory is shared between all modules. 
+The inter-module communication (using [uORB](../middleware/uorb.md)) is based on shared memory.
+The whole PX4 middleware runs in a single address space, i.e. memory is shared between all modules.
 
-> **Info** The system is designed such that with minimal effort it would
-> be possible to run each module in separate address space (parts that would need
-> to be changed include `uORB`, `parameter interface`, `dataman` and `perf`).
+> **Info** The system is designed such that with minimal effort it would be possible to run each module in separate address space (parts that would need to be changed include `uORB`, `parameter interface`, `dataman` and `perf`).
 
 There are 2 different ways that a module can be executed:
-- **Tasks**: The module runs in its own task with its own stack and process priority
-  (this is the more common way). 
-- **Work queues**: The module runs on a shared task, meaning that it does not own a stack. 
-  Multiple tasks run on the same stack with a single priority per work queue.
+- **Tasks**: The module runs in its own task with its own stack and process priority.
+- **Work queues**: The module runs on a shared task, meaning that it does not own a stack.
+  Multiple modules can run on the same stack with a single priority per work queue.
 
-  A task is scheduled by specifying a fixed time in the future.
-  The advantage is that it uses less RAM, but the task is not allowed to sleep
-  or poll on a message.
+  A task is scheduled by specifying a fixed time in the future, or via uORB topic update callback.
```

Comment:
1. This is indented, which means that it looks like it belongs with the Work queue information. Is that the case?
2. It says "A task is scheduled ..." - do you mean a task or a work queue?

So I "think" you probably mean (Please confirm):
- Task 
  - runs in its own stack and priority.
  - It can sleep or poll on a message and do blocking IO.
- Work queue:
  - shares a stack.
  - shares a priority with other work queues
  - cannot  sleep or poll on a message or do blocking IO.
- long running tasks should run in own task or own work queue - is this because queues are cooperative? or if not, why?

Further questions:
- Above implies you can have multiple work queues? If so, how do you choose which one? How many items should there be on a queue?
- Are work queues cooperative? i.e. do they have to cede control to other queues? If so we should explain that.

Reply:
- I'd change `Work queue: ` -> `Work queue task: `, as mentioned above, the work queue itself specifies essentially the thread, on which the tasks/items/modules are running. Not sure what the best term is.

> long running tasks should run in own task or own work queue - is this because queues are cooperative? or if not, why?

Yes, and everything is serialized, so a work queue item cannot be interrupted by another item on the same work queue.

> Above implies you can have multiple work queues? If so, how do you choose which one?

Correct, you specify on initialization: https://github.com/PX4/Firmware/blob/master/src/modules/load_mon/load_mon.cpp#L121
Here are all of them: https://github.com/PX4/Firmware/blob/master/platforms/common/include/px4_platform_common/px4_work_queue/WorkQueueManager.hpp#L49.

> How many items should there be on a queue?

As many as you want - at some point you'll notice negative effects, but we're not there yet.

> Are work queues cooperative? i.e. do they have to cede control to other queues? If so we should explain that.

A work queue is essentially a task/thread: it can be interrupted by higher-priority tasks.

Reply:
Thanks @bkueng 

I couldn't see an answer for this bit below. Can you check it is correct? Note modified in bits to be more clear

1. This is indented, which means that it looks like it belongs with the Work queue information. Is that the case?
2. It says "A task is scheduled ..." - do you mean a task or a work queue?

So I "think" you probably mean (Please confirm):
- Task 
  - runs in its own stack and priority.
  - It can sleep or poll on a message and do blocking IO.
- Work queue task:
  - shares a stack with other work queue tasks.
  - shares a priority with other work queue tasks on the queue
  - cannot  sleep or poll on a message or do blocking IO. Basically it gets its turn eventually (I assume it is round robin?)

Reply:
> This is indented, which means that it looks like it belongs with the Work queue information. Is that the case?

Correct

> It says "A task is scheduled ..." - do you mean a task or a work queue?

This is where we need clarity on the naming. It's the work queue task, but not the work queue thread itself.

> So I "think" you probably mean (Please confirm):

Confirmed.

> Basically it gets its turn eventually (I assume it is round robin?)

Yes added to a linked list. We might add priorities in future.